### PR TITLE
Jib sample

### PR DESCRIPTION
**Pull request by dgageot**

### CODE REVIEW

File: examples/annotated-skaffold.yaml [hunk 1]
```diff
@@ -61,6 +61,15 @@ build:
     # bazel:
     #  target: //:skaffold_example.tar
 
+    # jibMaven is an artifacts built with Maven, typically (but not only) a Java project.
```
Comment (briandealwis) [inline examples/annotated-skaffold.yaml]: nit: `artifacts` should be singular, or `is for artifacts`. Could I suggest the following?
```
# jibMaven builds containers using the Jib plugin for Maven
```

Do you see this file as being documentation?  Is it worth adding some detail about the `module` and `profile` items?

File: examples/annotated-skaffold.yaml [hunk 2]
```diff
@@ -61,6 +61,15 @@ build:
...
+    # jibMaven:
+    #  module: modulename
+    #  profile: profilename
+
+    # jibGradle is an artifacts built with Gradle, typically (but not only) a Java project.
```
Comment (briandealwis) [inline examples/annotated-skaffold.yaml]: And ditto re: `artifacts` or the suggested rewrite.

File: examples/jib/pom.xml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                <groupId>org.springframework.boot</groupId>
+                <artifactId>spring-boot-maven-plugin</artifactId>
+            </plugin>
+            <plugin>
+                <groupId>com.google.cloud.tools</groupId>
+                <artifactId>jib-maven-plugin</artifactId>
+                <version>0.9.11</version>
+                <configuration>
+                    <from>
+                        <image>gcr.io/distroless/java@sha256:b430543bea1d8326e767058bdab3a2482ea45f59d7af5c5c61334cd29ede88a1</image>
```
Comment (briandealwis) [inline examples/jib/pom.xml]: Why did you choose a specific tag, as opposed to `:latest`?

File: examples/annotated-skaffold.yaml (hunk 2 above)
Comment (dgageot) [reply examples/annotated-skaffold.yaml]: I'll rewrite it

**Comment by dgageot** [comment]
@briandealwis I took your feedback into account. `integration` is a necessary duplicate

File: examples/jib/README.adoc [hunk 4]
```diff
@@ -0,0 +1,31 @@
+=== Example: Jib
+:icons: font
+
+Jib is one of the supported builders in Skaffold.
+[Jib](https://github.com/GoogleContainerTools/jib) builds Docker and OCI images
+for your Java applications and is available as plugins for Maven and Gradle.
+
+The way you configure it in `skaffold.yaml` is the following build stanza:
```
Comment (briandealwis) [inline examples/jib/README.adoc]: I think it's worth mentioning that this example is for a standalone Maven project, where all dependencies are resolved from outside. The Jib builder requires that the projects are configured to use the Jib plugins for Maven or Gradle.  Multi-module builds require a bit additional configuration.

File: examples/jib/pom.xml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+                <artifactId>spring-boot-maven-plugin</artifactId>
+            </plugin>
+            <plugin>
+                <groupId>com.google.cloud.tools</groupId>
+                <artifactId>jib-maven-plugin</artifactId>
+                <version>0.9.11</version>
+                <configuration>
+                    <container>
+                        <jvmFlags>
+                            <jvmFlag>-Djava.security.egd=file:/dev/./urandom</jvmFlag>
```
[the remark this answers is not in the document]
Comment (dgageot) [reply examples/jib/pom.xml]: Because of a funny bug: https://bugs.java.com/bugdatabase/view_bug.do?bug_id=6202721

**Comment by anthonydahanne** [comment]
hello ! I've tried to run the samples 
```
git fetch dgageot
git checkout jib-sample
cd examples/jib
```
and then 

- I updated the references to "gcr.io/k8s-skaffold/skaffold-jib" with "anthonydahanne/test:skaffold"
- I had minikube running
- and I run ```skaffold dev```  but I ended up with : 
```FATA[0009] exiting dev mode because the first build failed: building [anthonydahanne/test:skaffold]: digest not found ```

Using debug logs, I had : 
```
[adah@MCADAH02 jib (jib-sample)]$ skaffold dev --verbosity debug
INFO[0000] Skaffold &{Version:v0.16.0 GitVersion: GitCommit:78e443973ee7475ee66d227431596351cf5e2caf GitTreeState:clean BuildDate:2018-10-11T23:05:05Z GoVersion:go1.11.1 Compiler:gc Platform:darwin/amd64} 
DEBU[0000] Defaulting build type to local build         
DEBU[0000] Defaulting deploy type to kubectl            
INFO[0000] Using kubectl context: minikube              
DEBU[0000] Using builder: local                         
DEBU[0000] Running command: [minikube docker-env --shell none] 
DEBU[0000] Command output: stdout DOCKER_TLS_VERIFY=1
DOCKER_HOST=tcp://192.168.99.100:2376
DOCKER_CERT_PATH=/Users/adah/.minikube/certs
DOCKER_API_VERSION=1.35
, stderr:  
DEBU[0000] push value not present, defaulting to false because localCluster is true 
DEBU[0000] Running command: [mvn jib:_skaffold-files -q] 
DEBU[0003] Command output: stdout /Users/adah/workspaces/skaffold/examples/jib/pom.xml
/Users/adah/workspaces/skaffold/examples/jib/src/main/java
/Users/adah/workspaces/skaffold/examples/jib/src/main/resources
/Users/adah/workspaces/skaffold/examples/jib/src/main/resources
/Users/adah/workspaces/skaffold/examples/jib/src/main/jib
, stderr:  
DEBU[0003] could not stat dependency: stat /Users/adah/workspaces/skaffold/examples/jib/src/main/resources: no such file or directory 
DEBU[0003] could not stat dependency: stat /Users/adah/workspaces/skaffold/examples/jib/src/main/resources: no such file or directory 
DEBU[0003] could not stat dependency: stat /Users/adah/workspaces/skaffold/examples/jib/src/main/jib: no such file or directory 
DEBU[0003] Found dependencies for jibMaven artifact: [/Users/adah/workspaces/skaffold/examples/jib/pom.xml /Users/adah/workspaces/skaffold/examples/jib/src/main/java /Users/adah/workspaces/skaffold/examples/jib/src/main/java/hello /Users/adah/workspaces/skaffold/examples/jib/src/main/java/hello/Application.java /Users/adah/workspaces/skaffold/examples/jib/src/main/java/hello/HelloController.java] 
Starting build...
Found [minikube] context, using local docker daemon.
Building [anthonydahanne/test:skaffold]...
INFO[0003] Building .: /usr/local/bin/mvn, [mvn prepare-package jib:dockerBuild -Dimage=jib__087e74d83f207d1cd326fc12cd7e0110bb02da61] 
DEBU[0003] Running command: [mvn prepare-package jib:dockerBuild -Dimage=jib__087e74d83f207d1cd326fc12cd7e0110bb02da61] 
[INFO] Scanning for projects...
[INFO] 
[INFO] -------------------< org.skaffold:hello-spring-boot >-------------------
[INFO] Building hello-spring-boot 0.1.0
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-resources-plugin:3.0.2:resources (default-resources) @ hello-spring-boot ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /Users/adah/workspaces/skaffold/examples/jib/src/main/resources
[INFO] skip non existing resourceDirectory /Users/adah/workspaces/skaffold/examples/jib/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.7.0:compile (default-compile) @ hello-spring-boot ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:3.0.2:testResources (default-testResources) @ hello-spring-boot ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /Users/adah/workspaces/skaffold/examples/jib/src/test/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.7.0:testCompile (default-testCompile) @ hello-spring-boot ---
[INFO] No sources to compile
[INFO] 
[INFO] --- maven-surefire-plugin:2.21.0:test (default-test) @ hello-spring-boot ---
[INFO] No tests to run.
[INFO] 
[INFO] --- jib-maven-plugin:0.9.11:dockerBuild (default-cli) @ hello-spring-boot ---
[WARNING] Base image 'gcr.io/distroless/java' does not use a specific image digest - build may not be reproducible
[INFO] 
[INFO] Containerizing application to Docker daemon as jib__087e74d83f207d1cd326fc12cd7e0110bb02da61...
[INFO] 
[INFO] Getting base image gcr.io/distroless/java...
[INFO] Building dependencies layer...
[INFO] Building classes layer...
[INFO] Finalizing...
[INFO] Loading to Docker daemon...
[INFO] 
[INFO] Container entrypoint set to [java, -Djava.security.egd=file:/dev/./urandom, -cp, /app/resources:/app/classes:/app/libs/*, hello.Application]
[INFO] 
[INFO] Built image to Docker daemon as jib__087e74d83f207d1cd326fc12cd7e0110bb02da61
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 4.897 s
[INFO] Finished at: 2018-10-14T09:46:07-04:00
[INFO] ------------------------------------------------------------------------
DEBU[0009] Defaulting build type to local build         
DEBU[0009] Defaulting deploy type to kubectl            
INFO[0009] Using kubectl context: minikube              
DEBU[0009] Using builder: local                         
DEBU[0009] push value not present, defaulting to false because localCluster is true 
Cleaning up...
DEBU[0009] manifests apiVersion: v1
kind: Service
metadata:
  name: web
spec:
  ports:
  - port: 8080
    name: http
  type: LoadBalancer
  selector:
    app: web
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: web
spec:
  selector:
    matchLabels:
      app: web
  template:
    metadata:
      labels:
        app: web
    spec:
      containers:
      - name: web
        image: anthonydahanne/test:skaffold 
DEBU[0009] Running command: [kubectl --context minikube delete --ignore-not-found=true -f -] 
Cleanup complete in 222.996062ms
FATA[0009] exiting dev mode because the first build failed: building [anthonydahanne/test:skaffold]: digest not found 
```

**Comment by anthonydahanne** [comment]
ok, so I found and fixed my mistakes : 
* Skaffold and Jib were using my Docker for Mac docker daemon, instead of using my Minikube own docker daemon; hence the "digest not found"; to fix that, I had to run, just before executing skaffold : 

``` eval $(minikube docker-env)```

* Then I hit a second issue : 
```
FATA[0013] exiting dev mode because the first build failed: building [anthonydahanne/test:skaffold]: tagging: Error parsing reference: "anthonydahanne/test:skaffold:7e79003b-dirty-b43b301" is not a valid repository/tag: invalid reference format
```
To fix that one, I just needed to remove the tag after my image name(replacing ```anthonydahanne/test:skaffold``` with ```anthonydahanne/test``` ), since, I should have known , skaffold and jib themselves tag the image

So I got it working and I could display a brilliant "Hello Anthony" after updating the HelloController :-)

**Comment by anthonydahanne** [comment]
I have 1 question though : I've seen in the source code, that's it's possible to append profiles, using, for example :
```
    jibMaven:
      profile: "prometheus"
```
Can we also pass -D vm arguments as well ?
Thanks ! (let me know if I'm abusing this Github issue, I'll post my questions somewhere more adequate !)

**Comment by briandealwis** [comment]
Not at the moment. You can define Maven properties in a profile. And we don't actually validate the profile value so you should be able to specify multiple profile IDs by separating them with commas.

What are you trying to do?

**Comment by anthonydahanne** [comment]
> What are you trying to do?

```-DskipTests```

That would not be convenient to create a profile for each VM argument or their combinations...

**Comment by dgageot** [comment]
Thank you @anthonydahanne for the feedback. I'll create a ticket for the image name.

**Comment by briandealwis** [comment]
One possible workaround (of sorts) is to use environment variables.  You can also activate a profile with an environment variable.

It may make sense for Skaffold to expose some information as to the build mode.  For example, that `skaffold dev` is in use so don't run tests.

File: examples/annotated-skaffold.yaml [hunk 6]
```diff
@@ -61,6 +61,15 @@ build:
     # bazel:
     #  target: //:skaffold_example.tar
 
+    # jibMaven builds containers using the Jib plugin for Maven.
+    # jibMaven:
+    #  module: modulename     # selects which maven module to build
```
Comment (coollog) [inline examples/annotated-skaffold.yaml]: Should we leave this out for now? (Since it's currently unimplemented)

File: examples/jib/pom.xml (hunk 5 above)
Comment (coollog) [inline examples/jib/pom.xml]: Perhaps we should also just include the other container-enabling JVM flags here too:

```
				  <jvmFlag>-XX:+UnlockExperimentalVMOptions</jvmFlag>
				  <jvmFlag>-XX:+UseCGroupMemoryLimitForHeap</jvmFlag>
```

**Comment by dgageot** [comment]
@GoogleContainerTools/java-tools should be ready for review

**Comment by dgageot** [comment]
Something is maybe flaky with kaniko local mode

**Pull request merged**